src/cmd/ksh93/tests/signal.sh

```diff
@@ -151,7 +154,7 @@ actual=$(echo $(
 ))
 [[ $actual == $expect ]] ||
     log_error "subshell ignoring signal does not send signal to parent" "$expect" "$actual"
-read -u9 -t0.01 x || log_error 'parent unexpectedly consumed fifo go signal'
+read -u9 -t0.1 x || log_error 'parent unexpectedly consumed fifo go signal'
```

Comment:
I have no idea why I worded that error message like that. If this `read` fails, due to timing out, it means the subcommand unexpectedly executed its `read x < fifo9` which isn't supposed to happen. Please replace `parent` with `subcommand`.

Reply:
If increasing the timeout fixes the problem it means the alarm signal was processed and interrupted the read syscall and thus the read command. It means your system was slow enough that the kernel delivered the signal before the read was started. This contradicts the comment you added to `sigalrm()`. Did you really determine `sigalrm()` isn't getting invoked or was that just your best guess as to what was going wrong?

In fact, I can induce a hang at this point in the test by simply adding `sh_delay(0.01);` to `sh_readline()` after the `sh_timeradd()` call. So I'm okay with increasing the timeout to make the test more robust on heavily loaded systems. But I think the comment you added to *src/cmd/ksh93/sh/timers.c* should be removed.

Reply:
Got it. I misread the source comment. And I've confirmed via debug printf()'s that the problem is essentially as you say. What's interesting is that when the failure occurs `sigalrm()` is being run three orders of magnitude sooner than it should be run. On my system approximately 10 usec after `sh_timeradd()` returns rather than the expected 10,000 usec. Even more interesting sometimes a second SIGALRM is delivered while the read syscall is in progress.

The right thing to do is rewrite this to use `select()` or `poll()` rather than `alarm()` and `siglongjmp()`.